grep: Implement '-n', the flag for outputting line numbers

BenWiederhake opened the pull request:
I always write `grep -Ern` by default without thinking about it, so I tripped up a lot while doing #10793. Let's improve our grep!

![Bildschirmfoto_2021-11-06_12-58-13](https://user-images.githubusercontent.com/2690845/140608766-a6940f82-f82e-41e4-826b-9167b4b4a1f7.png)

The color choice is totally arbitrary, because we don't follow GNU grep anyway.

EDIT: You might need to try the regex `=true$`, because #10793 already got merged.
EDIT2: Fixed the screenshot

BenWiederhake commented:
Changes since last push:
- Fix the long-name `-n`
- Export the manpage (detected by #10655)

BertalanD commented on Base/usr/share/man/man1/grep.md:
```diff
@@ -16,6 +16,7 @@ $ grep [--recursive] [--extended-regexp] [--regexp Pattern] [-i] [--invert-match
 * `-E`, `--extended-regexp`: Extended regular expressions
 * `-e Pattern`, `--regexp Pattern`: Pattern
 * `-i`: Make matches case-insensitive
+* `-n`, `--invert-match`: Output line-numbers
```
```suggestion
* `-n`, `--line-numbers`: Output line-numbers
```

nico commented on Base/usr/share/man/man1/grep.md:
````diff
@@ -5,7 +5,7 @@ grep
 
 ```sh
-$ grep [--recursive] [--extended-regexp] [--regexp Pattern] [-i] [--invert-match] [--quiet] [--no-messages] [--binary-mode ] [--text] [-I] [--color WHEN] [file...]
+$ grep [--recursive] [--extended-regexp] [--regexp Pattern] [-i] [--invert-match] [--invert-match] [--quiet] [--no-messages] [--binary-mode ] [--text] [-I] [--color WHEN] [file...]
````
```suggestion
$ grep [--recursive] [--extended-regexp] [--regexp Pattern] [-i] [--invert-match] [--line-numbers] [--quiet] [--no-messages] [--binary-mode ] [--text] [-I] [--color WHEN] [file...]
```

Also, should we maybe sort the flags up here (?) Also also, do intentionally only put the long forms up here?

BenWiederhake replied:
Regarding the copypasto: Will fix that

Regarding sorting: Meh, I don't really see the point.

Regarding long form: This file is [auto-generated through ArgsParser](https://github.com/SerenityOS/serenity/blob/066b1183cdd7dd0bd62bf0546cda7068946afcf2/Userland/Libraries/LibCore/ArgsParser.cpp#L267), and that code was mostly copied from the terminal formatter. Sure, that makes sense, but it's out of scope for this PR.

Now you know which code to change :)

bgianfo commented on Userland/Utilities/grep.cpp:
```diff
@@ -168,11 +172,11 @@ int main(int argc, char** argv)
             }
 
-            while (file->can_read_line()) {
+            for (size_t line_number = 0; file->can_read_line(); ++line_number) {
```
Shouldn't this start at 1?

BenWiederhake replied:
Argh! Good catch. I was so fixated with the correctness of the *other* loop, that I totally forgot to check this one.

BenWiederhake commented:
Changes since last push:
- Fixed the silly off-by-one if the data doesn't come from stdin
- Rebased for convenience
- Updated the screenshot in the first post
- Made extra sure that line numbers are correct for both stdin and `-r` and ``

bgianfo commented:
LGTM!
